packages/techdocs-common/src/stages/publish/googleStorage.ts

```diff
@@ -0,0 +1,222 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import path from 'path';
+import express from 'express';
+import walk from 'klaw';
+import { Storage, UploadResponse } from '@google-cloud/storage';
+import { Logger } from 'winston';
+import { Entity, EntityName } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { getHeadersForFileExtension, supportedFileType } from './helpers';
+import { PublisherBase, PublishRequest } from './types';
+
+export class GoogleGCSPublish implements PublisherBase {
+  static fromConfig(config: Config, logger: Logger): PublisherBase {
+    let pathToKey = '';
+    let projectId = '';
+    let bucketName = '';
+    try {
+      pathToKey = config.getString('techdocs.publisher.google.pathToKey');
+      projectId = config.getString('techdocs.publisher.google.projectId');
+      bucketName = config.getString('techdocs.publisher.google.bucketName');
+    } catch (error) {
+      throw new Error(
+        "Since techdocs.publisher.type is set to 'google_gcs' in your app config, " +
+          'pathToKey, projectId and bucketName are required in techdocs.publisher.google ' +
+          'required to authenticate with Google Cloud Storage.',
+      );
+    }
+
+    const storageClient = new Storage({
+      projectId: projectId,
+      keyFilename: pathToKey,
+    });
+
+    // Check if the defined bucket exists. Being able to connect means the configuration is good
+    // and the storage client will work.
+    storageClient
+      .bucket(bucketName)
+      .getMetadata()
+      .then(() => {
+        logger.info(
+          `Successfully connected to the GCS bucket ${bucketName} in the GCP project ${projectId}.`,
+        );
+      })
+      .catch(reason => {
+        logger.error(
+          `Could not retrieve metadata about the GCS bucket ${bucketName} in the GCP project ${projectId}. ` +
+            'Make sure the GCP project and the bucket exists and the access key located at the path ' +
+            "techdocs.publisher.google.pathToKey defined in app config has the role 'Storage Object Creator'. " +
+            'Refer to https://backstage.io/docs/features/techdocs/using-cloud-storage',
+        );
+        throw new Error(`from GCS client library: ${reason.message}`);
+      });
+
+    return new GoogleGCSPublish(storageClient, bucketName, logger);
+  }
+
+  constructor(
+    private readonly storageClient: Storage,
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the GCS bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<{}> {
```

Comment:
i think you wanted `Promise<void>` here, then you can `resolve(undefined)` below